### PR TITLE
Additionally Vectorize string.IndexOfAny for value lengths 2,3,4,5

### PR DESCRIPTION
**Pull request by benaadams**
Also for `where T : char`
```csharp
IndexOfAny<T>(this Span<T> span, T value0, T value1)
IndexOfAny<T>(this Span<T> span, T value0, T value1, T value2)
IndexOfAny<T>(this Span<T> span, ReadOnlySpan<T> values)
IndexOfAny<T>(this ReadOnlySpan<T> span, T value0, T value1)
IndexOfAny<T>(this ReadOnlySpan<T> span, T value0, T value1, T value2)
IndexOfAny<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> values)
```

Extends Vectorization for 1 value to 2, 3, 4 and 5 value searches

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/String.Searching.cs [hunk 1]
```diff
@@ -105,11 +105,13 @@ public int IndexOfAny(char[] anyOf, int startIndex, int count)
             if (anyOf.Length == 2)
             {
                 // Very common optimization for directory separators (/, \), quotes (", '), brackets, etc
-                return IndexOfAny(anyOf[0], anyOf[1], startIndex, count);
+                var result = SpanHelpers.IndexOfAny(ref Unsafe.Add(ref _firstChar, startIndex), anyOf[0], anyOf[1], count);
```
Comment (benaadams) [inline src/System.Private.CoreLib/shared/System/String.Searching.cs]: Isn't currently a way to combing the length check for this branch `anyOf.Length == 2` with bounds check elimination for the accesses `anyOf[0], anyOf[1]`?

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 2]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+                        goto Found2;
+                    if (pCh[3] == value0 || pCh[3] == value1 || pCh[3] == value2)
+                        goto Found3;
+
+                    pCh += 4;
+                }
+
+                while (length > 0)
+                {
+                    length -= 1;
```
Comment (krwq) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: length--;

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 3]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+                    // This is equivalent to:
+                    //         unaligned = ((int)pCh % Unsafe.SizeOf<Vector<ushort>>()) / elementsPerByte
+                    //         length = (Vector<ushort>.Count - unaligned) % Vector<ushort>.Count
+                    const int elementsPerByte = sizeof(ushort) / sizeof(byte);
+                    int unaligned = ((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) / elementsPerByte;
+                    length = (Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1);
+                }
+            SequentialScan:
+#endif
+                while (length >= 4)
```
Comment (krwq) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: won't the SequentialScan be run regardless of IsHardwareAccelerated? Is JIT skipping this code path?

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 4]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
             }
         }
 
+        public static unsafe int IndexOfAny(ref char searchSpace, char value0, char value1, int length)
```
Comment (krwq) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: Do we have test coverage for corner cases this code might affect?

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 5]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+        {
+            Debug.Assert(length >= 0);
+
+            fixed (char* pChars = &searchSpace)
+            {
+                char* pCh = pChars;
+                char* pEndCh = pCh + length;
+
+#if !netstandard11
+                if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: Out of scope for this PR, but it might be nice to try this with Hardware Intrinsics as well.

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 6]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+                {
+                    // Figure out how many characters to read sequentially until we are vector aligned
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: Why start with a sequential read?

You should be able to do a single unaligned read, mask out bits that would be covered by the first aligned read, and then do the vectorized check (and I'm not even sure you need to mask for this algorithm).

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 7]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+                    pCh += 1;
+                }
+#if !netstandard11
+                // We get past SequentialScan only if IsHardwareAccelerated is true. However, we still have the redundant check to allow
+                // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+                if (Vector.IsHardwareAccelerated && pCh < pEndCh)
+                {
+                    // Get the highest multiple of Vector<ushort>.Count that is within the search space.
+                    // That will be how many times we iterate in the loop below.
+                    // This is equivalent to: length = Vector<ushort>.Count * ((int)(pEndCh - pCh) / Vector<ushort>.Count)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: I really wish the the JIT did this optimization.

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs (hunk 6 above)
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: I've just copied the `IndexOf<char>(this Span<char> span, char value0)` method and used the same multi-comparison with vectorized `BitwiseOr` as `byte` to widen it to 2 and 3 chars.

So an improved approach to vectorization, whether an unaligned start or using instrinsics directly should also apply to those.

Might be worth a tracking issue?

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: > Might be worth a tracking issue?

Probably. I'll log one sometime Tuesday, if I don't get to it before then.

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs (hunk 4 above)
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: via string.IndexOfAny but I have added testing via the Span entry points which should give fuller coverage https://github.com/dotnet/corefx/pull/32069

**Comment by benaadams** [comment]
@danmosemsft while testing this I encounted a lot of use via `FileSystemName` and 5 chars        
```csharp
private static readonly char[] s_wildcardChars =
{
    '\"', '<', '>', '*', '?'
};
```

Wondering if this should be extended for just 2-3 to go up to 5, rather than moving to a loop per item at 4+? 

@tannergooding would there be enough xmm registers?

**Comment by tannergooding** [comment]
x64 likely has enough registers, but you'd want to look at the codegen in any case, since the effectiveness of the various registers likely depends on surrounding code and the platform abi (which would define, for example, if the registers have to be saved by the caller, before transferring to the callee)

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs [hunk 8]
```diff
@@ -175,6 +175,206 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
...
+                while (length > 0)
+                {
+                    length -= 1;
+
+                    if (pCh[0] == value0 || pCh[0] == value1)
+                        goto Found;
+
+                    pCh += 1;
+                }
+#if !netstandard11
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: We don't need these ifdefs anymore. Keeo the code path though cc @jkotas

Comment (ahsonkhan) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: See https://github.com/dotnet/coreclr/pull/19863#discussion_r216022805

**Comment by tannergooding** [comment]
We are good to merge this now, correct?

**Comment by benaadams** [comment]
@dotnet-bot  test

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
```
MESSAGE:
Assert.Throws() Failure\nExpected: typeof(System.OverflowException)\nActual: (No exception was thrown)
+++++++++++++++++++
STACK TRACE:
at System.Runtime.InteropServices.Tests.SafeBufferTests.Initialize_NumBytesTimesSizeOfEachElement_ThrowsOverflowException() in /mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/_/fx/src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/SafeBufferTests.cs:line 32
```

**Comment by benaadams** [comment]
```
Assert failure(PID 5308 [0x000014bc], Thread: 8072 [0x1f88]): 
CONTRACT VIOLATION by GetValueTypeForGUID at "d:\j\workspace\x64_checked_w---d7295605\src\vm\interoputil.cpp" @ 2862
```

**Comment by benaadams** [comment]
@tannergooding think so

**Comment by benaadams** [comment]
Added tests for 4,5 values https://github.com/dotnet/corefx/pull/32734 so they can be also vectorized

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs (hunk 8 above)
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: Removed

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.cs [hunk 9]
```diff
@@ -491,6 +510,13 @@ public static int IndexOfAny<T>(this ReadOnlySpan<T> span, T value0, T value1, T
                     Unsafe.As<T, byte>(ref value1),
                     Unsafe.As<T, byte>(ref value2),
                     span.Length);
+            if (typeof(T) == typeof(char))
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.cs]: Changed

**Comment by benaadams** [comment]
Added vectorization for 4 and 5 values

File: src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs (hunk 6 above)
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs]: Not sure it would help. The vectorization is fast at discounting characters; but not fast at then finding a match where the sequential by itself would probably be faster - rather than doing an overlapping vector at start and overlapping vector at the end.

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.cs [hunk 10]
```diff
@@ -589,6 +671,61 @@ public static int IndexOfAny<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> value
                     ref Unsafe.As<T, byte>(ref MemoryMarshal.GetReference(values)),
                     values.Length);
 
+            if (typeof(T) == typeof(char))
+            {
+                ref var valueRef = ref Unsafe.As<T, char>(ref MemoryMarshal.GetReference(values));
+                if (values.Length == 5)
+                {
+                    // Length 5 is a common length for FileSystemName expression (", <, >, *, ?) and in preference to 2 as it has an explicit overload
```
Comment (benaadams) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.cs]: @JeremyKuhne helpful to have this first? `FileSystemName` switches between two different length arrays 5 and 2

**Comment by benaadams** [comment]
OSX failing on files
```
Test Result (5 failures / +5)
file_io.GetFileSize.test1
file_io.GetFileSizeEx.test1
file_io.SetFilePointer.test5
file_io.SetFilePointer.test6
file_io.SetFilePointer.test7
```

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.cs (hunk 10 above)
Comment (JeremyKuhne) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.cs]: Sorry, missed this. I hope it becomes less common for 5 in the future, but for now it is by far and away the most common case. The new default when using `EnumerationOptions` is the simple type (i.e. 2).

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by benaadams** [comment]
Ah, corefx tests are failing due to https://github.com/dotnet/coreclr/issues/20405

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
Windows_NT x64 Checked corefx_baseline issue is unrelated and previously reported https://github.com/dotnet/coreclr/issues/20373
```
Thread: 5960 [0x1748]): CONTRACT VIOLATION by GetValueTypeForGUID 
at "d:\j\workspace\x64_checked_w---d7295605\src\vm\interoputil.cpp" @ 2862

MODE_COOPERATIVE encountered while thread is in preemptive state.
```

**Comment by benaadams** [comment]
> We are good to merge this now, correct?

@tannergooding yes, tests are now in corefx dotnet/corefx#32734 and pass with these changes. Feedback has been addressed.

**Comment by tannergooding** [comment]
Thanks for the work @benaadams
